### DISCUSSION TITLE
The pep484 of functions where () and filter () is invalid for 2.0b1

### DISCUSSION THREAD

**Discussion by wu-clan** [sqlalchemy/sqlalchemy#8638] [Usage Questions] [answered]
I am trying to fully upgrade the existing project to sqlalchemy-2.0, because the arrival of PEP484(happy). When using the where() or filter() method, the follow -up prompt is invalid. I don't know how to test, but I checked filter_by(), it can work normally. At the same time, I saw that they were pointing to the type "SelfDMLWhereBase".

where():

![where()](https://user-images.githubusercontent.com/52145145/195982552-02998c4c-a1bd-4fa2-aee1-d56957bb09cb.png)

filter()

![filter()](https://user-images.githubusercontent.com/52145145/195982585-a1f9e4c7-73f4-40f3-9bfc-357285627628.png)

filter_by():

![filter_by()](https://user-images.githubusercontent.com/52145145/195982626-4bbf31a4-e9b2-48ae-8c17-a48999141010.png)

**Comment by CaselIT** [comment]
Hi,

I'm afraid you will need to provide a bit more information.
What's the ide? Have you uninstalled all the strub packages?
Looking at this this seems to be an ide issue 

Does mypy or pyright give an error?

**Reply by zzzeek** [reply]
it looks like vscode, I'll try now

**Reply by wu-clan** [reply]
> it looks like vscode, I'll try now

In fact, it's pycharm professional-2022.2.3, the theme deceives you, haha

**Comment by wu-clan** [comment]
The doubt is that the filter_by() can correctly provide a prompt, but where() did they not work, I created a new virtual environment and cleaned the IDE cache. The IDE did not give a warning, but the method prompts did not seem to be very good Inspection, maybe a screenshot of the correct prompt is needed to eliminate my doubts.

**Reply by zzzeek** [reply]
cleaning a cache is not sufficient, IDEs have typeshed stubs hardcoded in their config directories and may have to be deleted manually (and also for each time you upgrade the IDE).

**Reply by wu-clan** [reply]
This is a great help, thank you

**Comment by zzzeek** [comment, marked as answer]
Here's what I have for vscode.

it looks like you have a different but similar IDE but it seems unlikely that it doesnt know how to read types correctly, so you'd need to see that typeshed isn't embedded in your config somewhere.  IDEs will have typeshed stubs in random places.

A way to confirm the typing in a way we can agree on is to use reveal_type(), then run mypy on the file.

![Screenshot from 2022-10-15 08-40-03](https://user-images.githubusercontent.com/128223/195987133-206b8132-ceef-4fe2-b02c-3eafc7284cbc.png)
![Screenshot from 2022-10-15 08-40-38](https://user-images.githubusercontent.com/128223/195987134-06275380-340b-461e-beae-39744236d0a8.png)
![Screenshot from 2022-10-15 08-41-35](https://user-images.githubusercontent.com/128223/195987137-8d1687d6-543a-4dab-adea-fa89aaab0c1b.png)

**Reply by zzzeek** [reply]
```py
from __future__ import annotations

from sqlalchemy import update
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import Session


class Base(DeclarativeBase):
    pass


class User(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[str]


s = Session()

name = "x"

u1 = update(User).filter_by(data=name)
u2 = update(User).filter(User.data == name)

reveal_type(u1)
reveal_type(u2)

```

mypy output

```
$ mypy test3.py 
test3.py:28: note: Revealed type is "sqlalchemy.sql.dml.Update"
test3.py:29: note: Revealed type is "sqlalchemy.sql.dml.Update"
Success: no issues found in 1 source file
```

**Comment by wu-clan** [comment]
Thank you very much, I check it now

**Comment by wu-clan** [comment]
Tag reminder, after removing all .pyi stubs, confirming that the database models are using `Mapped`
